server/backend/testspg/crud/test_crud_role.py

```diff
@@ -0,0 +1,218 @@
+from testspg.constants import TEST_NOT_EXIST_ROLE_ID, \
+    TEST_FOM_DEV_REVIEWER_ROLE_ID, \
+    TEST_FOM_DEV_SUBMITTER_ROLE_ID, \
+    TEST_FOM_DEV_APPLICATION_ID, \
+    TEST_CREATOR, \
+    TEST_FOM_TEST_APPLICATION_ID, \
+    TEST_NOT_EXIST_APPLICATION_ID
+
+LOGGER = logging.getLogger(__name__)
+
+TEST_NEW_ROLE = "TEST_READ"
+TEST_NEW_ROLE_TWO = "TEST_WRITE"
+TEST_FOREST_CLIENT_NUMBER = "00000002"
+TEST_FOREST_CLIENT_NUMBER_TWO = "00000003"
+TEST_ROLE_PURPOSE = "test role"
+TEST_ROLE_CREATE = {
+    "application_id": TEST_FOM_DEV_APPLICATION_ID,
+    "role_name": TEST_NEW_ROLE,
+    "role_purpose": TEST_ROLE_PURPOSE,
+    "create_user": TEST_CREATOR,
+    "role_type_code": constants.RoleType.ROLE_TYPE_CONCRETE,
+}
+
+
+def test_get_roles(db_pg_connection: Session):
+    roles = crud_role.get_roles(db_pg_connection)
+    assert len(roles) > 1
+
+
+def test_get_role(db_pg_connection: Session):
+    # get non exists role
+    found_role = crud_role.get_role(db_pg_connection, TEST_NOT_EXIST_ROLE_ID)
+    assert found_role is None
+
+    # get concreate role
+    found_role = crud_role.get_role(db_pg_connection, TEST_FOM_DEV_REVIEWER_ROLE_ID)
+    assert found_role.role_id == TEST_FOM_DEV_REVIEWER_ROLE_ID
+    assert found_role.role_name == "FOM_REVIEWER"
+    assert found_role.application_id == TEST_FOM_DEV_APPLICATION_ID
+
+    # get abstract role
+    found_role = crud_role.get_role(db_pg_connection, TEST_FOM_DEV_SUBMITTER_ROLE_ID)
+    assert found_role.role_id == TEST_FOM_DEV_SUBMITTER_ROLE_ID
+    assert found_role.role_name == "FOM_SUBMITTER"
+    assert found_role.application_id == TEST_FOM_DEV_APPLICATION_ID
+
+
+def test_create_role(db_pg_connection: Session):
+    new_role = crud_role.create_role(
+        schemas.FamRoleCreate(**TEST_ROLE_CREATE),
+        db_pg_connection,
+    )
+    assert new_role.role_name == TEST_NEW_ROLE
+
+    # verify role created
+    found_role = crud_role.get_role(db_pg_connection, new_role.role_id)
+    assert found_role.role_id == new_role.role_id
+    assert found_role.role_name == TEST_NEW_ROLE
+    assert found_role.application_id == TEST_FOM_DEV_APPLICATION_ID
+    assert found_role.role_type_code == constants.RoleType.ROLE_TYPE_CONCRETE
+
+
+def test_create_role_duplicate(db_pg_connection: Session):
+    # can not create the role with same role name, even with a different creator
+    with pytest.raises(IntegrityError) as e:
+        crud_role.create_role(
+            schemas.FamRoleCreate(
+                **{
+                    "application_id": TEST_FOM_DEV_APPLICATION_ID,
+                    "role_name": TEST_NEW_ROLE,
+                    "role_purpose": TEST_ROLE_PURPOSE,
+                    "create_user": "ANOTHER_CREATOR",
+                    "role_type_code": constants.RoleType.ROLE_TYPE_CONCRETE,
+                }
+            ),
+            db_pg_connection,
+        )
+    assert str(e.value).find("duplicate key value violates unique constraint") != -1
```

Comment:
Hi Ian, yes this one is relied on the previous one "test_create_role". We don't have a delete role method, so for role tests I didn't write any cleanup.

Reply:
Hi Catherine, 
Previously in old "**tests**" there were some **fixtures** setup that **cleanup** test data after each function.
That might did the job previously ("tests" folder), that's one idea we can borrow; or I think in this case,
although there is no "**delete role**" written at 'crud' or 'router', we can still use **db session** with 
direct sql delete as well.

However, maybe there are better way. I am thinking we might need to change our foundation "**pgtests**" setup to **rollback** after each test (**NOT** destroy/recreate db) so we can write less fixture to cleanup. 

Also, if I checkout this pr branch and run in terminal for just one test file (under "testpg" folder) I should
be able to run tests individually and successful. But using these tow commands and compare you will see the failure:
"`pytest --ignore tests testspg/crud/test_crud_application.py`" => passed
"`pytest --ignore tests testspg/router/test_router_application.py`" => failed.

Reply:
I added the cleanup for every tests

Reply:
For the reason why run this test individually it failed "pytest --ignore tests testspg/router/test_router_application.py" => failed, it's because for router tests, we didn't call the method to setup the pg container directly. When we run all the tests together, the pg container will be called by other tests, and will be setup ready. But when run the router tests individually, we missed that part. 

I added the "db_pg_container" in the router tests where we missed that.  In fact, we only need to add it to the first test method in each router test file, it will only setup once, but to reduce confusion, I added to all, so it looks like:
<img width="220" alt="Screen Shot 2023-04-11 at 4 12 00 PM" src="https://user-images.githubusercontent.com/23131735/231308027-2511e0b8-5b26-462f-82dd-ac9d63b0c3a3.png">

Reply:
> I added the cleanup for every tests

Thank you Catherine!
Although I was thinking something else, but what you did is good enough I think. Thanks!

Reply:
> For the reason why run this test individually it failed "`pytest --ignore tests testspg/router/test_router_application.py`" => failed it's because for router tests, we didn't call the method to setup the pg container directly. When we run all the tests together, the pg container will be called by other tests, and will be setup ready. But when run the router tests individually, we missed that part.
> 
> I added the "db_pg_container" in the router tests where we missed that. In fact, we only need to add it to the first test method in each router test file, it will only setup once, and will be shut down once all tests done. But to reduce confusion, I added to all router tests that missed it, so it looks like: <img alt="Screen Shot 2023-04-11 at 4 12 00 PM" width="220" src="https://user-images.githubusercontent.com/23131735/231308027-2511e0b8-5b26-462f-82dd-ac9d63b0c3a3.png">

For Pytest, adding fixture into each function when needed is fine (but not just at the first one) because Pytest will determine when to run the fixture by its scope. In this case, your db_pg_container is in scope="session" (as intended) then it is good. If this test file is written in another style in Pytest (like 'grouping' or have 'setup' before each test) then it can be more obvious to developers the container session is up and ready before all tests.